[quant][graphmode][fx] e2e test with torchvision

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43534 [quant][graphmode][fx][fix] enable per channel quantization for functional ops
* **#43526 [quant][graphmode][fx] e2e test with torchvision**

Summary:
Add e2e tests for fx graph mode quantization on torchvision and make sure it matches
current eager mode quantization

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23306683](https://our.internmc.facebook.com/intern/diff/D23306683)

jerryzh168 committed:
[quant][graphmode][fx] Testing torchvision

Summary:
Add tests for graph mode quantization on torchvision and make sure it matches
current eager mode quantization

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -757,3 +768,177 @@ def forward(self, x):
+        model_list = quantized_model_list
+        # slice need to be fixed in symbolic tracing(https://github.com/pytorch/pytorch/issues/43511)
+        model_list = set(model_list) - {'googlenet', 'inception_v3'}
+        # getattr should not be used as node name(https://github.com/pytorch/pytorch/issues/43522)
```
cc @suo

jerryzh168 committed:
Update on "[quant][graphmode][fx] Testing torchvision"


Summary:
Add tests for graph mode quantization on torchvision and make sure it matches
current eager mode quantization

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/fx/graph_module.py:
```diff
@@ -41,6 +41,7 @@ class GraphModuleImpl(cls):  # type: ignore
     def __init__(self, root: torch.nn.Module, graph: Graph):
         super().__init__()
         self.root = root
+        self.training = self.root.training
```
can I make this change here for now to unblock this PR? @jamesr66a @suo

jamesr66a replied:
Where is this actually used? It shouldn't matter if the top-level module has `training` set; the only thing that this affects is stuff like batch norm. Where are we actually depending on this flag?

jerryzh168 replied:
people might be relying on the training attribute of the graph module, e.g. https://github.com/pytorch/pytorch/blob/master/test/quantization/test_quantize_fx.py#L873

vkuzo commented on test/quantization/test_quantize_fx.py:
```diff
@@ -757,3 +768,177 @@ def forward(self, x):
+class TestQuantizeFxModels(QuantizationTestCase):
+    @skipIfNoTorchVision
+    def test_torchvision(self):
+        from torchvision import models
```
are there cross-repo compatibility concerns to think through here?  AFAIK torchvision doesn't fully enforce a way to make breaking API changes in a BC/FC compatible way (it's more of an honor system where the promise it that it is guaranteed on major releases, but not on every PR).

jerryzh168 replied:
I don't feel we are relying on something that will change in torchvision if that is what you are worrying about

jerryzh168 replied:
I also saw in https://github.com/pytorch/pytorch/blob/master/.circleci/config.yml#L1825 that it is using
```
pip install torchvision
```
so it should just be installing major releases

vkuzo replied:
sg, thanks for explaining!

vkuzo commented on test/quantization/test_quantize_fx.py:
```diff
@@ -757,3 +768,177 @@ def forward(self, x):
+            quantizer = Quantizer()
+            prepared = quantizer.prepare(graph_module, qconfig_dict)
+
+            if mode == 'ddp':
+                mp.spawn(run_ddp,
+                         args=(world_size, prepared),
+                         nprocs=world_size,
+                         join=True)
```
I feel like for DDP we can test that it runs without errors on a toy model, and deduce the correctness of real models from the guarantees of FX and DDP provide.  So, not sure if we need to test torchvision in DDP specifically as a part of the test suite, although I'm flexible

jerryzh168 replied:
this is e2e test and we have a reference implementation of eager mode so I think it is valuable.

jerryzh168 replied:
we can also have unit tests for ddp to just make it runs as well

vkuzo replied:
sure, would it be ok to separate DDP testing into its own file then, so that the reader can ignore it unless they are specifically working on it

I'm not sure if we need to test all the models with DDP either, testing one model which contains SyncBatchNorm + a backwards pass + the guarantees of DDP will be enough to verify all models are correct if one is correct

jerryzh168 replied:
I can make ddp just test one model in this test case, splitting it to a separate file would probably mean copying the other code and I think we should avoid that

vkuzo commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -101,6 +106,99 @@ def test_only_train_fn(model, train_data, loss_fn=_default_loss_fn):
+          .format(top1=top1, top5=top5))
+    return
+
+def ddp_setup(rank, world_size):
```
there should be some utilities in `test/distributed` for this, can we reuse?

jerryzh168 replied:
you mean https://github.com/pytorch/pytorch/blob/master/test/distributed/test_distributed.py? it looks like they are doing everything inline, I didn't see a function. also I'm not sure how to reuse this https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_distributed.py#L207

vkuzo replied:
np, can we split this into a separate file then?  Ideally people who look at `common_quantization.py` shouldn't have to read code related to DDP.

jerryzh168 replied:
Sure

jerryzh168 replied:
you mean split this to common_quantization_ddp or something?

jerryzh168 replied:
I think common_quantization just contains all utility functions related to quantization tests, looking at other common_* code, i don't feel we should create a new file just for this

vkuzo replied:
I accepted the PR, I trust you.  I do feel like DDP testing ideally should be fully separate, and we shouldn't be testing every single thing with DDP but instead making sure that DDP team supports fx properly, and trusting their system to work.  It's a high level question of do we test A*B*C things (test everything, but more expensive to maintain), or do we test A*B things and B*C things with clean and well defined APIs and trust that A*B*C works by induction from A*B and B*C working.

vkuzo replied:
> I think common_quantization just contains all utility functions related to quantization tests, looking at other common_* code, i don't feel we should create a new file just for this

also an optional suggestion.  Usually after some LOC (~200-400LOC), it's harder for a code reader to fully understand the entire file quickly.  Keeping files small and single purposed allows for easier code browsing.

jerryzh168 replied:
I agree we don't need to test everything, but unit test and integration tests are both important, in this case we are testing resnet18 with ddp and it is a integration test that makes sure ddp + quantization + fx works, this can not be tested in unit test only I think.

re code readability, I agree we don't want to mix unrelated things in a single file. but common_quantization is more of a util file where we put all util functions in. I think we can split the models to a separate file in common_quantization.py. but the other code are utility functions used in test.

vkuzo commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -176,6 +274,13 @@ def wrapper(*args, **kwargs):
             fn(*args, **kwargs)
     return wrapper
 
+try:
```
thoughts on keeping the torchvision tests and utils in separate files?  It seems to be for a different use case (domain testing of correctness), and it would be useful for code readers to be able to ignore it unless they are specifically working on torchvision

jerryzh168 replied:
this is not really about torchvision I think. we are using torchvision because it is already quantized with eager mode. the goal is to do end to end testing.

jerryzh168 replied:
the name of the PR might be confusing, I'll change it

raghuramank100 replied:
One option is to keep this test as part of the torchvision repo

vkuzo replied:
yeah, my high level thoughts are:
1. it's definitely valuable to know that torchvision eager mode and FX produce the same results
2. I'm not sure if checking in a torchvision e2e test would be a net benefit, I'd be worried about this test being flaky and I'm not sure if depending on torchvision from unit tests is good
3. I think we can get the same guarantees with less testing (we can assume DDP is working correctly without testing it here, and we can assume that if all the building blocks and patterns of torchvision models work, then the models themselves would work).
4. I don't feel super strongly about (3) and I could be wrong as well, we could check in this code and if it leads to problems later we can delete it

vkuzo replied:
> the goal is to do end to end testing.

optional, but feels like e2e comparisons of small building blocks could give us the same guarantees, without the torchvision dependency

jerryzh168 replied:
fx is already depending on torchvision, I think it's fine: https://github.com/pytorch/pytorch/blob/master/test/test_fx.py#L10-L15

it's just easier to test torchvision directly. if we want to test sub blocks of torchvision models we will also need to write the test for them. we could of course spend time to do that, but I don't think it's a priority at the moment and not sure how valuable it is to remove the torchvision dependency

vkuzo commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -101,6 +106,99 @@ def test_only_train_fn(model, train_data, loss_fn=_default_loss_fn):
+
+def train_one_epoch(model, criterion, optimizer, data_loader, device, ntrain_batches):
+    model.train()
+    top1 = AverageMeter('Acc@1', ':6.2f')
+    top5 = AverageMeter('Acc@5', ':6.2f')
+    avgloss = AverageMeter('Loss', '1.5f')
```
nit: can we delete these, since we are comparing the model outputs and not the accuracy?

jerryzh168 replied:
sure

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -757,3 +768,177 @@ def forward(self, x):
+
+        # mobilenet: dropout error RuntimeError: "bernoulli_scalar_cpu_" not implemented for 'QUInt8'
+        # incpetion_v3: looks like there is some problem with AuxLogits
+        quantized_not_working = [('qat', 'mobilenet_v2'),
```
mobilenet_v2 should work with QAT. Why is this excluded here?

jerryzh168 replied:
not sure about this, currently mobilenet_v2 can't traced anyway, will revisit when it is enabled.

jerryzh168 replied:
the error I saw previously is: RuntimeError: "bernoulli_scalar_cpu_" not implemented for 'QUInt8'

raghuramank100 replied:
That corresponds to drop out being called with quantized tensors, which is incorrect. Does this mean that we are missing a dequantize somehere?

jerryzh168 replied:
dropout can support quantized tensor right? I think it is probably we set training to true and it is not supported. It might be fixed now actually since currently I always make model eval before convert

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -757,3 +768,177 @@ def forward(self, x):
             expected_node_occurrence=count_check,
             expected_node_list=order_check)
+
+class TestQuantizeFxModels(QuantizationTestCase):
```
Should we have this test in torchvision repo instead? We had this test previously: https://github.com/pytorch/vision/blob/master/test/test_quantized_models.py

jerryzh168 replied:
does torchvision depend on pytorch master?

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -757,3 +768,177 @@ def forward(self, x):
+
+                # calibration
+                if mode == 'ddp':
+                    mp.spawn(run_ddp,
```
Will this run on all combinations of CPU/GPU and OS?

jerryzh168 replied:
no, will only test on multi gpu setting, I will split this to a separate test in the update

jerryzh168 committed:
Update on "[quant][graphmode][fx] e2e test with torchvision"


Summary:
Add e2e tests for fx graph mode quantization on torchvision and make sure it matches
current eager mode quantization

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23306683](https://our.internmc.facebook.com/intern/diff/D23306683)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] e2e test with torchvision"


Summary:
Add e2e tests for fx graph mode quantization on torchvision and make sure it matches
current eager mode quantization

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23306683](https://our.internmc.facebook.com/intern/diff/D23306683)

[ghstack-poisoned]
